crates/redeem/src/lib.rs

```diff
@@ -336,9 +341,103 @@ pub mod pallet {
             Self::_mint_tokens_for_reimbursed_redeem(vault_id, redeem_id)?;
             Ok(().into())
         }
+
+        #[pallet::weight(<T as Config>::WeightInfo::request_redeem())]
+        #[transactional]
+        pub fn self_redeem(
+            origin: OriginFor<T>,
+            currency_pair: DefaultVaultCurrencyPair<T>,
+            amount_wrapped: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let account_id = ensure_signed(origin)?;
+            let vault_id = VaultId::new(account_id, currency_pair.collateral, currency_pair.wrapped);
+            let amount_wrapped = Amount::new(amount_wrapped, vault_id.wrapped_currency());
+
+            self_redeem::execute::<T>(vault_id, amount_wrapped)?;
+
+            Ok(().into())
+        }
     }
 }
 
+mod self_redeem {
+    use super::*;
+
+    pub(crate) fn execute<T: Config>(vault_id: DefaultVaultId<T>, amount_wrapped: Amount<T>) -> DispatchResult {
+        // ensure that vault is not liquidated and not banned
+        ext::vault_registry::ensure_not_banned::<T>(&vault_id)?;
+
+        // for self-redeem, dustAmount is effectively 1 satoshi
+        ensure!(!amount_wrapped.is_zero(), Error::<T>::AmountBelowDustAmount);
+
+        let (fees, consumed_issued_tokens) = calculate_token_amounts::<T>(&vault_id, &amount_wrapped)?;
+
+        take_user_tokens::<T>(&vault_id.account_id, &consumed_issued_tokens, &fees)?;
+
+        update_vault_tokens::<T>(&vault_id, &consumed_issued_tokens)?;
+
+        Pallet::<T>::deposit_event(Event::<T>::SelfRedeem {
+            vault_id,
+            amount: consumed_issued_tokens.amount(),
+            fee: fees.amount(),
+        });
+
+        Ok(())
+    }
```

Comment:
My original code looked surprisingly complex for such a simple function, so this was a bit of an experiment to attempt to make the code easier to read. I like how clear this function is now, although I did end up paying the price in increased verbosity.